File: Stage19/C/Problem24060.c
#include <stdio.h>
#include <stdlib.h>

int A[500000], tmp[500000];
int K, count = 0, result = -1;

void merge(int left, int mid, int right) {
    int i = left;
    int j = mid + 1;
    int t = left;

    while (i <= mid && j <= right) {
        if (A[i] <= A[j]) tmp[t++] = A[i++];
        else tmp[t++] = A[j++];
        count++;
        if (count == K) result = tmp[t - 1];
    }

    while (i <= mid) {
        tmp[t++] = A[i++];
        count++;
        if (count == K) result = tmp[t - 1];
    }

    while (j <= right) {
        tmp[t++] = A[j++];
        count++;
        if (count == K) result = tmp[t - 1];
    }

    for (int x = left; x <= right; x++)
        A[x] = tmp[x];
}

void mergeSort(int left, int right) {
    if (left < right) {
        int mid = (left + right) / 2;
        mergeSort(left, mid);
        mergeSort(mid + 1, right);
        merge(left, mid, right);
    }
}

int main(void) {
    int N;
    scanf("%d %d", &N, &K);

    for (int i = 0; i < N; i++)
        scanf("%d", &A[i]);

    mergeSort(0, N - 1);
    printf("%d\n", result);

    return 0;
}

File: Stage19/Python/Problem24060.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

def merge_sort(left, right):
    if left < right:
        mid = (left + right) // 2
        merge_sort(left, mid)
        merge_sort(mid + 1, right)
        merge(left, mid, right)

def merge(left, mid, right):
    global A, tmp, K, count, result

    i, j = left, mid + 1
    t = left

    while i <= mid and j <= right:
        if A[i] <= A[j]:
            tmp[t] = A[i]
            i += 1
        else:
            tmp[t] = A[j]
            j += 1
        t += 1
        count += 1
        if count == K:
            result = tmp[t - 1]

    while i <= mid:
        tmp[t] = A[i]
        i += 1
        t += 1
        count += 1
        if count == K:
            result = tmp[t - 1]

    while j <= right:
        tmp[t] = A[j]
        j += 1
        t += 1
        count += 1
        if count == K:
            result = tmp[t - 1]

    for x in range(left, right + 1):
        A[x] = tmp[x]


def main():
    global A, tmp, K, count, result
    N, K = map(int, input().split())
    A = list(map(int, input().split()))
    tmp = [0] * N
    count = 0
    result = -1

    merge_sort(0, N - 1)
    print(result)


if __name__ == "__main__":
    main()